src/main/java/org/ow2/proactive/connector/iaas/cloud/provider/jclouds/openstack/OpenstackJCloudsProvider.java

```diff
@@ -191,33 +200,45 @@ public void removeInstancePublicIp(Infrastructure infrastructure, String instanc
 
     @Override
     public SimpleImmutableEntry<String, String> createKeyPair(Infrastructure infrastructure, Instance instance) {
-        throw new UnsupportedOperationException();
+        String keyPairName = KEY_PAIR_PREFIX + "-" + UUID.randomUUID();
+        NovaApi novaApi = buildNovaApi(infrastructure);
+        KeyPair keyPair = novaApi.getKeyPairApi(region).get().create(keyPairName);
+
+        log.info("Openstack key-pair created: " + keyPair.getName() + " [" + keyPair.toString() + "]");
+        return new SimpleImmutableEntry<>(keyPair.getName(), keyPair.toString());
     }
 
     private void validatePlateformOperation(NovaApi novaApi) {
-        if (!novaApi.getFloatingIPApi(REGION).isPresent()) {
-            throw new NotSupportedException("Operation not supported for this Openstack cloud");
+        if (!novaApi.getFloatingIPApi(region).isPresent()) {
+            throw new NotSupportedException("Operation not supported by the targeted Openstack version");
         }
     }
 
     private Server createOpenstackInstance(Instance instance, ServerApi serverApi, CreateServerOptions serverOptions) {
+        openstackUtil.validateOpenstackInstanceParameters(instance);
         ServerCreated serverCreated = serverApi.create(instance.getTag(),
                                                        instance.getImage(),
                                                        instance.getHardware().getType(),
                                                        serverOptions);
 
+        log.info("Server instance created: " + serverCreated.toString());
+
         return serverApi.get(serverCreated.getId());
     }
 
     private final Instance createInstanceFromNode(Server server) {
-        return Instance.builder()
-                       .id(REGION + "/" + server.getId())
-                       .tag(server.getName())
-                       .image(server.getImage().getName())
-                       .number("1")
-                       .hardware(Hardware.builder().type(server.getFlavor().getName()).build())
-                       .status(server.getStatus().name())
-                       .build();
+        Instance instance = Instance.builder()
+                                    .id(region + "/" + server.getId())
+                                    .tag(server.getName())
+                                    .image(server.getImage().getName())
+                                    .number("1")
```

Comment:
Can you create a final variable in this method to describe what is `1` and use it here.

Reply:
Done